src/api/app/models/workflow.rb

```diff
@@ -7,18 +7,57 @@ def initialize(workflow:, scm_extractor_payload:, token:)
     @token = token
   end
 
+  def valid?
+    !unsupported_steps? && !invalid_steps?
+  end
+
   def steps
-    steps = []
-    return steps if @workflow['steps'].blank?
+    return [] if @workflow['steps'].blank?
+    return [] if unsupported_steps?
 
-    @workflow['steps'].each do |step|
-      step.each do |step_name, step_instructions|
-        next if SUPPORTED_STEPS[step_name].blank?
+    extract_workflow_steps
+  end
 
-        new_step = SUPPORTED_STEPS[step_name].new(step_instructions: step_instructions, scm_extractor_payload: @scm_extractor_payload, token: @token)
-        steps << new_step if new_step.allowed_event_and_action?
+  def errors
+    unsupported_steps.each_with_object([]) do |step_definition, acc|
+      step_definition.each do |step_name, _|
+        acc << "'#{step_name}' is not a supported step."
       end
+      acc
+    end
+  end
+
+  private
+
+  def unsupported_steps?
+    unsupported_steps.any?
+  end
+
+  def invalid_steps?
+    extract_workflow_steps.reject(&:valid?).any?
+  end
+
+  def unsupported_steps
+    @workflow['steps'].each_with_object([]) do |step_definition, acc|
```

Comment:
We should memoize the result of this method since it's used in multiple methods throughout this file.

Something like `@unsupported_steps ||= (the whole method code)`

Reply:
Yeah, this is something I know I have to improve, but I haven't, yet.

Reply:
I've made some improvements in a separate commit